File: packages/ccode/src/config/keywords.ts
/**
 * Agent Keywords Configuration Loader.
 *
 * Loads trigger keywords from configuration files:
 * - Default: packages/ccode/src/agent/keywords.default.json
 * - User: ~/.codecoder/keywords.json
 *
 * User configuration is merged on top of defaults, allowing selective overrides.
 */

import path from "path"
import os from "os"
import { parse as parseJsonc, type ParseError as JsoncParseError, printParseErrorCode } from "jsonc-parser"
import { mergeDeep } from "remeda"
import { Log } from "@/util/log"

// Import default keywords configuration
import defaultKeywords from "../agent/keywords.default.json"

const log = Log.create({ service: "keywords-loader" })

// ─────────────────────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Trigger rule for agent detection.
 */
export type TriggerRule = string | AdvancedTriggerRule

/**
 * Advanced trigger rule with type and options.
 */
export interface AdvancedTriggerRule {
  type: "keyword" | "pattern" | "context" | "event"
  value: string
  priority?: number
  description?: string
}

/**
 * Keywords configuration for a single agent.
 */
export interface AgentKeywords {
  triggers: TriggerRule[]
  aliases: string[]
  priority: number
  enabled?: boolean
}

/**
 * Default agent settings.
 */
export interface DefaultsConfig {
  agent: string
  cli_agent: string
  im_agent: string
  use_implicit_matching: boolean
}

/**
 * Root keywords configuration.
 */
export interface KeywordsConfig {
  version: string
  agents: Record<string, AgentKeywords>
  defaults: DefaultsConfig
}

// ─────────────────────────────────────────────────────────────────────────────
// Configuration Paths
// ─────────────────────────────────────────────────────────────────────────────

/** User configuration directory */
const CONFIG_DIR = path.join(os.homedir(), ".codecoder")

/** User keywords configuration file */
const KEYWORDS_PATH = path.join(CONFIG_DIR, "keywords.json")

// ─────────────────────────────────────────────────────────────────────────────
// Loading Functions
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Load a JSON/JSONC file.
 * Returns null if file doesn't exist or parsing fails.
 */
async function loadJsonFile<T>(filepath: string): Promise<T | null> {
  try {
    const text = await Bun.file(filepath).text()
    const errors: JsoncParseError[] = []
    const data = parseJsonc(text, errors, { allowTrailingComma: true })

    if (errors.length) {
      const errorDetails = errors.map((e) => printParseErrorCode(e.error)).join(", ")
      log.warn("JSONC parse errors", { path: filepath, errors: errorDetails })
      return null
    }

    return data as T
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") {
      return null
    }
    log.error("Failed to load config file", { path: filepath, error })
    return null
  }
}

/**
 * Load keywords configuration.
 *
 * Priority (lowest to highest):
 * 1. Default keywords (bundled)
 * 2. User keywords (~/.codecoder/keywords.json)
 */
export async function loadKeywords(): Promise<KeywordsConfig> {
  // Start with default configuration
  let config = defaultKeywords as KeywordsConfig

  // Try to load user configuration
  const userConfig = await loadJsonFile<Partial<KeywordsConfig>>(KEYWORDS_PATH)

  if (userConfig) {
    // Deep merge user config on top of defaults
    config = mergeDeep(config, userConfig) as KeywordsConfig
    log.debug("Loaded user keywords config", {
      path: KEYWORDS_PATH,
      agents: Object.keys(userConfig.agents ?? {}).length,
    })
  }

  return config
}

// ─────────────────────────────────────────────────────────────────────────────
// Cached Configuration
// ─────────────────────────────────────────────────────────────────────────────

/** Cached keywords configuration */
let cachedConfig: KeywordsConfig | null = null

/**
 * Get keywords configuration (cached).
 *
 * Loads configuration on first call and caches it for subsequent calls.
 */
export async function getKeywords(): Promise<KeywordsConfig> {
  if (!cachedConfig) {
    cachedConfig = await loadKeywords()
  }
  return cachedConfig
}

/**
 * Get keywords configuration synchronously (uses default if not loaded).
 *
 * Call `getKeywords()` first to ensure user config is loaded.
 */
export function getKeywordsSync(): KeywordsConfig {
  return cachedConfig ?? (defaultKeywords as KeywordsConfig)
}

/**
 * Reload keywords configuration (clears cache).
 */
export async function reloadKeywords(): Promise<KeywordsConfig> {
  cachedConfig = null
  return getKeywords()
}

// ─────────────────────────────────────────────────────────────────────────────
// Detection Functions
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Detect agent from @mention alias.
 *
 * @param message - User message to analyze
 * @param config - Keywords configuration
 * @returns Agent name if found, null otherwise
 *
 * @example
 * ```ts
 * detectAlias("@macro 分析PMI", config) // => "macro"
 * detectAlias("@宏观 分析PMI", config) // => "macro"
 * ```
 */
export function detectAlias(message: string, config: KeywordsConfig): string | null {
  const text = message.trim()

  // Must start with @
  if (!text.startsWith("@")) {
    return null
  }

  // Extract the mention (first word after @)
  const parts = text.split(/\s+/)
  const mention = parts[0]?.slice(1).toLowerCase()
  if (!mention) {
    return null
  }

  // Search through all agents for matching alias
  for (const [agentName, keywords] of Object.entries(config.agents)) {
    if (keywords.enabled === false) {
      continue
    }

    // Check exact agent name match
    if (agentName.toLowerCase() === mention) {
      return agentName
    }

    // Check aliases
    for (const alias of keywords.aliases) {
      if (alias.toLowerCase() === mention) {
        return agentName
      }
    }
  }

  return null
}

/**
 * Detect agent from message content using keyword triggers.
 *
 * This performs implicit keyword matching. Note: disabled by default
 * to avoid misrouting.
 *
 * @param message - User message to analyze
 * @param config - Keywords configuration
 * @returns Agent name with highest priority match, or null
 */
export function detectTrigger(message: string, config: KeywordsConfig): string | null {
  // Check if implicit matching is enabled
  if (!config.defaults.use_implicit_matching) {
    return null
  }

  const text = message.toLowerCase()
  let bestMatch: { agent: string; priority: number } | null = null

  for (const [agentName, keywords] of Object.entries(config.agents)) {
    if (keywords.enabled === false) {
      continue
    }

    for (const trigger of keywords.triggers) {
      let matched = false
      let priority = keywords.priority

      if (typeof trigger === "string") {
        matched = text.includes(trigger.toLowerCase())
      } else {
        switch (trigger.type) {
          case "keyword":
            matched = text.includes(trigger.value.toLowerCase())
            break
          case "pattern":
          case "context":
            try {
              matched = new RegExp(trigger.value, "i").test(text)
            } catch {
              matched = false
            }
            break
          case "event":
            // Events are handled separately
            matched = false
            break
        }
        priority = trigger.priority ?? keywords.priority
      }

      if (matched) {
        if (!bestMatch || priority > bestMatch.priority) {
          bestMatch = { agent: agentName, priority }
        }
        break // Only count first match per agent
      }
    }
  }

  return bestMatch?.agent ?? null
}

/**
 * Combined agent detection: alias → trigger → default.
 *
 * @param message - User message to analyze
 * @param defaultAgent - Fallback agent if no match found
 * @param config - Keywords configuration
 * @returns Detected agent name (never returns null)
 */
export function detectAgent(
  message: string,
  defaultAgent: string,
  config: KeywordsConfig
): string {
  // 1. Try @mention alias detection
  const aliasMatch = detectAlias(message, config)
  if (aliasMatch) {
    return aliasMatch
  }

  // 2. Try implicit keyword matching (if enabled)
  const triggerMatch = detectTrigger(message, config)
  if (triggerMatch) {
    return triggerMatch
  }

  // 3. Fallback to default
  return defaultAgent
}

/**
 * Get the trigger priority for a specific keyword.
 *
 * @param keyword - The keyword to look up
 * @param agentName - The agent to search in
 * @param config - Keywords configuration
 * @returns Priority value, or default (5) if not found
 */
export function getTriggerPriority(
  keyword: string,
  agentName: string,
  config: KeywordsConfig
): number {
  const agent = config.agents[agentName]
  if (!agent) {
    return 5
  }

  const lowerKeyword = keyword.toLowerCase()
  for (const trigger of agent.triggers) {
    if (typeof trigger === "string") {
      if (trigger.toLowerCase() === lowerKeyword) {
        return agent.priority
      }
    } else if (trigger.value.toLowerCase() === lowerKeyword) {
      return trigger.priority ?? agent.priority
    }
  }

  return agent.priority
}

File: services/zero-core/src/common/keywords.rs
//! Agent keywords configuration for routing.
//!
//! This module loads agent trigger keywords from a configuration file
//! (`~/.codecoder/keywords.json`) and provides functions for detecting
//! which agent should handle a given input.
//!
//! # Configuration File
//!
//! ```json
//! {
//!   "agents": {
//!     "macro": {
//!       "triggers": ["macro", "GDP", "economy"],
//!       "aliases": ["macro", "宏观"],
//!       "priority": 6
//!     }
//!   },
//!   "defaults": {
//!     "agent": "general",
//!     "cli_agent": "build",
//!     "im_agent": "autonomous"
//!   }
//! }
//! ```
//!
//! # Usage
//!
//! ```rust,ignore
//! use zero_common::keywords::{detect_alias, load_keywords};
//!
//! let keywords = load_keywords();
//! if let Some(agent) = detect_alias("@macro 分析PMI", &keywords) {
//!     println!("Routing to agent: {}", agent);
//! }
//! ```

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::sync::OnceLock;

use super::config::config_dir;

// ============================================================================
// Types
// ============================================================================

/// Trigger rule for agent detection.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum TriggerRule {
    /// Simple keyword string
    Simple(String),
    /// Advanced rule with type and options
    Advanced {
        #[serde(rename = "type")]
        trigger_type: TriggerType,
        value: String,
        #[serde(default = "default_priority")]
        priority: u8,
        #[serde(default)]
        description: Option<String>,
    },
}

/// Type of trigger rule.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum TriggerType {
    /// Keyword contains match
    Keyword,
    /// Regex pattern match
    Pattern,
    /// Semantic context match
    Context,
    /// System event trigger
    Event,
}

/// Keywords configuration for a single agent.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentKeywords {
    /// Keywords that trigger this agent
    #[serde(default)]
    pub triggers: Vec<TriggerRule>,

    /// Aliases for @mention routing
    #[serde(default)]
    pub aliases: Vec<String>,

    /// Agent priority (1-10, higher = more preferred)
    #[serde(default = "default_priority")]
    pub priority: u8,

    /// Whether this agent's keywords are active
    #[serde(default = "default_enabled")]
    pub enabled: bool,
}

/// Default agent settings.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DefaultsConfig {
    /// Default agent when no triggers match
    #[serde(default = "default_agent")]
    pub agent: String,

    /// Default agent for CLI channel
    #[serde(default = "default_cli_agent")]
    pub cli_agent: String,

    /// Default agent for IM channels
    #[serde(default = "default_im_agent")]
    pub im_agent: String,

    /// Enable implicit keyword matching
    #[serde(default)]
    pub use_implicit_matching: bool,
}

impl Default for DefaultsConfig {
    fn default() -> Self {
        Self {
            agent: default_agent(),
            cli_agent: default_cli_agent(),
            im_agent: default_im_agent(),
            use_implicit_matching: false,
        }
    }
}

/// Root keywords configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KeywordsConfig {
    /// Configuration version
    #[serde(default = "default_version")]
    pub version: String,

    /// Agent keywords keyed by agent name
    #[serde(default)]
    pub agents: HashMap<String, AgentKeywords>,

    /// Default settings
    #[serde(default)]
    pub defaults: DefaultsConfig,
}

impl Default for KeywordsConfig {
    fn default() -> Self {
        Self {
            version: default_version(),
            agents: HashMap::new(),
            defaults: DefaultsConfig::default(),
        }
    }
}

// ============================================================================
// Default Value Functions
// ============================================================================

fn default_priority() -> u8 {
    5
}

fn default_enabled() -> bool {
    true
}

fn default_agent() -> String {
    "general".to_string()
}

fn default_cli_agent() -> String {
    "build".to_string()
}

fn default_im_agent() -> String {
    "autonomous".to_string()
}

fn default_version() -> String {
    "1.0.0".to_string()
}

// ============================================================================
// Loading Functions
// ============================================================================

/// Path to the keywords configuration file.
pub fn keywords_path() -> PathBuf {
    config_dir().join("keywords.json")
}

/// Load keywords configuration from file.
///
/// Returns Ok(config) if successful, Ok(default) if file doesn't exist.
pub fn load_keywords_from_file() -> KeywordsConfig {
    let path = keywords_path();

    if !path.exists() {
        tracing::debug!("Keywords config not found, using defaults");
        return KeywordsConfig::default();
    }

    match fs::read_to_string(&path) {
        Ok(content) => match serde_json::from_str::<KeywordsConfig>(&content) {
            Ok(config) => {
                tracing::info!(
                    path = %path.display(),
                    agents = config.agents.len(),
                    "Loaded keywords configuration"
                );
                config
            }
            Err(e) => {
                tracing::warn!(
                    path = %path.display(),
                    error = %e,
                    "Failed to parse keywords config, using defaults"
                );
                KeywordsConfig::default()
            }
        },
        Err(e) => {
            tracing::warn!(
                path = %path.display(),
                error = %e,
                "Failed to read keywords config, using defaults"
            );
            KeywordsConfig::default()
        }
    }
}

/// Global keywords configuration (lazy loaded).
static KEYWORDS: OnceLock<KeywordsConfig> = OnceLock::new();

/// Get the global keywords configuration.
pub fn keywords() -> &'static KeywordsConfig {
    KEYWORDS.get_or_init(load_keywords_from_file)
}

/// Load keywords configuration (for API compatibility).
pub fn load_keywords() -> &'static KeywordsConfig {
    keywords()
}

// ============================================================================
// Detection Functions
// ============================================================================

/// Detect agent from @mention alias.
///
/// Checks if the message starts with @<alias> and returns the matching agent name.
///
/// # Examples
///
/// ```rust,ignore
/// let result = detect_alias("@macro 分析PMI", keywords());
/// assert_eq!(result, Some("macro"));
///
/// let result = detect_alias("@宏观 分析PMI", keywords());
/// assert_eq!(result, Some("macro"));
/// ```
pub fn detect_alias<'a>(message: &str, config: &'a KeywordsConfig) -> Option<&'a str> {
    let text = message.trim();

    // Must start with @
    if !text.starts_with('@') {
        return None;
    }

    // Extract the mention (first word after @)
    let mention = text
        .split_whitespace()
        .next()?
        .trim_start_matches('@')
        .to_lowercase();

    // Search through all agents for matching alias
    for (agent_name, keywords) in &config.agents {
        if !keywords.enabled {
            continue;
        }

        // Check exact agent name match
        if agent_name.to_lowercase() == mention {
            return Some(agent_name.as_str());
        }

        // Check aliases
        for alias in &keywords.aliases {
            if alias.to_lowercase() == mention {
                return Some(agent_name.as_str());
            }
        }
    }

    None
}

/// Detect agent from message content using keyword triggers.
///
/// This performs implicit keyword matching (e.g., "macro" in "分析宏观经济").
/// Note: This is disabled by default to avoid misrouting.
///
/// Returns the agent name with highest priority match, or None.
pub fn detect_trigger<'a>(message: &str, config: &'a KeywordsConfig) -> Option<&'a str> {
    // Check if implicit matching is enabled
    if !config.defaults.use_implicit_matching {
        return None;
    }

    let text = message.to_lowercase();
    let mut best_match: Option<(&str, u8)> = None;

    for (agent_name, keywords) in &config.agents {
        if !keywords.enabled {
            continue;
        }

        for trigger in &keywords.triggers {
            let (matched, priority) = match trigger {
                TriggerRule::Simple(keyword) => {
                    (text.contains(&keyword.to_lowercase()), keywords.priority)
                }
                TriggerRule::Advanced {
                    trigger_type,
                    value,
                    priority,
                    ..
                } => {
                    let matched = match trigger_type {
                        TriggerType::Keyword => text.contains(&value.to_lowercase()),
                        TriggerType::Pattern => {
                            regex::Regex::new(value)
                                .map(|re| re.is_match(&text))
                                .unwrap_or(false)
                        }
                        TriggerType::Context => {
                            regex::Regex::new(value)
                                .map(|re| re.is_match(&text))
                                .unwrap_or(false)
                        }
                        TriggerType::Event => false, // Events are handled separately
                    };
                    (matched, *priority)
                }
            };

            if matched {
                match best_match {
                    Some((_, best_priority)) if priority > best_priority => {
                        best_match = Some((agent_name.as_str(), priority));
                    }
                    None => {
                        best_match = Some((agent_name.as_str(), priority));
                    }
                    _ => {}
                }
                break; // Only count first match per agent
            }
        }
    }

    best_match.map(|(name, _)| name)
}

/// Combined agent detection: alias → trigger → default.
///
/// # Arguments
///
/// * `message` - The user message to analyze
/// * `default_agent` - Fallback agent if no match found
/// * `config` - Keywords configuration to use
///
/// # Returns
///
/// The detected agent name (never returns None, falls back to default).
pub fn detect_agent<'a>(
    message: &str,
    default_agent: &'a str,
    config: &KeywordsConfig,
) -> &'a str {
    // 1. Try @mention alias detection
    if let Some(agent) = detect_alias(message, config) {
        // SAFETY: We're returning a reference to a string that lives in config,
        // but we need to return &'a str. Since config is 'static (from KEYWORDS),
        // this is safe, but we need to handle the lifetime correctly.
        // For now, we use the leaked string approach or just return as-is.
        return Box::leak(agent.to_string().into_boxed_str());
    }

    // 2. Try implicit keyword matching (if enabled)
    if let Some(agent) = detect_trigger(message, config) {
        return Box::leak(agent.to_string().into_boxed_str());
    }

    // 3. Fallback to default
    default_agent
}

// ============================================================================
// Tests
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    fn test_config() -> KeywordsConfig {
        let mut config = KeywordsConfig::default();

        config.agents.insert(
            "macro".to_string(),
            AgentKeywords {
                triggers: vec![
                    TriggerRule::Simple("macro".to_string()),
                    TriggerRule::Simple("GDP".to_string()),
                ],
                aliases: vec!["macro".to_string(), "宏观".to_string()],
                priority: 6,
                enabled: true,
            },
        );

        config.agents.insert(
            "trader".to_string(),
            AgentKeywords {
                triggers: vec![TriggerRule::Simple("trading".to_string())],
                aliases: vec!["trader".to_string(), "交易".to_string()],
                priority: 6,
                enabled: true,
            },
        );

        config.agents.insert(
            "plan".to_string(),
            AgentKeywords {
                triggers: vec![TriggerRule::Advanced {
                    trigger_type: TriggerType::Keyword,
                    value: "plan".to_string(),
                    priority: 10,
                    description: None,
                }],
                aliases: vec!["plan".to_string(), "planner".to_string()],
                priority: 8,
                enabled: true,
            },
        );

        config
    }

    #[test]
    fn test_detect_alias_english() {
        let config = test_config();
        assert_eq!(detect_alias("@macro 分析PMI", &config), Some("macro"));
        assert_eq!(detect_alias("@trader 技术分析", &config), Some("trader"));
        assert_eq!(detect_alias("@plan 设计架构", &config), Some("plan"));
    }

    #[test]
    fn test_detect_alias_chinese() {
        let config = test_config();
        assert_eq!(detect_alias("@宏观 分析PMI", &config), Some("macro"));
        assert_eq!(detect_alias("@交易 技术分析", &config), Some("trader"));
    }

    #[test]
    fn test_detect_alias_case_insensitive() {
        let config = test_config();
        assert_eq!(detect_alias("@MACRO 分析PMI", &config), Some("macro"));
        assert_eq!(detect_alias("@Macro 分析PMI", &config), Some("macro"));
    }

    #[test]
    fn test_detect_alias_no_match() {
        let config = test_config();
        assert_eq!(detect_alias("@unknown 什么", &config), None);
        assert_eq!(detect_alias("没有@mention", &config), None);
        assert_eq!(detect_alias("纯文本消息", &config), None);
    }

    #[test]
    fn test_detect_trigger_disabled() {
        let config = test_config();
        // Implicit matching is disabled by default
        assert_eq!(detect_trigger("分析宏观经济", &config), None);
    }

    #[test]
    fn test_detect_trigger_enabled() {
        let mut config = test_config();
        config.defaults.use_implicit_matching = true;

        assert_eq!(detect_trigger("分析macro形势", &config), Some("macro"));
        assert_eq!(detect_trigger("GDP增长数据", &config), Some("macro"));
    }

    #[test]
    fn test_detect_agent_fallback() {
        let config = test_config();
        let result = detect_agent("普通消息", "build", &config);
        assert_eq!(result, "build");
    }

    #[test]
    fn test_detect_agent_with_alias() {
        let config = test_config();
        let result = detect_agent("@macro 分析PMI", "build", &config);
        assert_eq!(result, "macro");
    }
}
